From 56ca5515aac22cddfbabdf2ef8ce5debf619903c Mon Sep 17 00:00:00 2001
From: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Date: Sat, 4 Mar 2023 11:48:05 +0530
Subject: [PATCH 2/4] mac80211: restructure vif and link conf for mlo mbssid
 support

For MLD, if Transmitting interface is MLD then tx link inside that MLD
has to be saved inside non-tx MLD's link.
Currently mbssid_tx_vif is inside vif structure which won't suffice for
finding tx link for any non-tx link.
Move mbssid_tx_vif to ieee80211_bss_conf and add mbssid_tx_vif_linkid to
the same to find TX vif and Tx link inside the TX vif for any non-tx
link.

Modify CSA completion and BCCA event notifier routines to fetch
non-tx vdev links based on mbssid_tx_vif and mbssid_tx_vif_linkid.

Signed-off-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
---
 include/net/mac80211.h |  4 +--
 net/mac80211/cfg.c     | 79 ++++++++++++++++++++++++++++++++++--------
 net/mac80211/iface.c   | 18 +++++++---
 3 files changed, 79 insertions(+), 22 deletions(-)

diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index d0497e9..0ce9009 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -739,6 +739,8 @@ struct ieee80211_bss_conf {
 	struct ieee80211_ftm_responder_params *ftmr_params;
 	/* Multiple BSSID data */
 	bool nontransmitted;
+	struct ieee80211_vif *mbssid_tx_vif;
+	int mbssid_tx_vif_linkid;
 	u8 transmitter_bssid[ETH_ALEN];
 	u8 bssid_index;
 	u8 bssid_indicator;
@@ -1963,8 +1965,6 @@ struct ieee80211_vif {
 	bool txqs_stopped[IEEE80211_NUM_ACS];
 	bool noqueue_enable;
 
-	struct ieee80211_vif *mbssid_tx_vif;
-
 #ifdef CPTCFG_MAC80211_PPE_SUPPORT
 	u32 ppe_vp_num;
 #endif
diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 0daa809..58ca5f4 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -142,8 +142,10 @@ static int ieee80211_set_ap_mbssid_options(struct ieee80211_sub_if_data *sdata,
 					   struct ieee80211_bss_conf *link_conf)
 {
 	struct ieee80211_sub_if_data *tx_sdata;
+	struct ieee80211_bss_conf *tx_link;
 
-	sdata->vif.mbssid_tx_vif = NULL;
+	link_conf->mbssid_tx_vif = NULL;
+	link_conf->mbssid_tx_vif_linkid = 0;
 	link_conf->bssid_index = 0;
 	link_conf->nontransmitted = false;
 	link_conf->ema_ap = false;
@@ -156,9 +158,18 @@ static int ieee80211_set_ap_mbssid_options(struct ieee80211_sub_if_data *sdata,
 		return -EINVAL;
 
 	if (tx_sdata == sdata) {
-		sdata->vif.mbssid_tx_vif = &sdata->vif;
+		link_conf->mbssid_tx_vif = &sdata->vif;
+		link_conf->mbssid_tx_vif_linkid = link_conf->link_id;
 	} else {
-		sdata->vif.mbssid_tx_vif = &tx_sdata->vif;
+		link_conf->mbssid_tx_vif = &tx_sdata->vif;
+		rcu_read_lock();
+		tx_link = rcu_dereference(tx_sdata->link[params.tx_link_id]);
+		if (!tx_link) {
+			rcu_read_unlock();
+			return -ENOLINK;
+		}
+		link_conf->mbssid_tx_vif_linkid = tx_link->link_id;
+		rcu_read_unlock();
 		link_conf->nontransmitted = true;
 		link_conf->bssid_index = params.index;
 	}
@@ -3676,23 +3687,41 @@ void ieee80211_csa_finish(struct ieee80211_vif *vif, unsigned int link_id)
 		return;
 	}
 
-	/* TODO: MBSSID with MLO changes */
-	if (vif->mbssid_tx_vif == vif) {
+	if (link->conf->mbssid_tx_vif == vif &&
+	    link->conf->mbssid_tx_vif_linkid == link->conf->link_id) {
 		/* Trigger ieee80211_csa_finish() on the non-transmitting
 		 * interfaces when channel switch is received on
 		 * transmitting interface
 		 */
+		unsigned int link_id_iter;
+		u16 valid_links;
 		struct ieee80211_sub_if_data *iter;
+		struct ieee80211_link_data *link_iter;
 
 		list_for_each_entry_rcu(iter, &local->interfaces, list) {
 			if (!ieee80211_sdata_running(iter))
 				continue;
 
-			if (iter == sdata || iter->vif.mbssid_tx_vif != vif)
+			if (iter == sdata)
 				continue;
 
-			ieee80211_queue_work(&iter->local->hw,
-					     &iter->deflink.csa_finalize_work);
+			/* check link 0 by default for Non-ML non-tx vif's deflinks */
+			valid_links = iter->vif.valid_links | BIT(0);
+			for_each_set_bit(link_id_iter, &valid_links,
+					 IEEE80211_MLD_MAX_NUM_LINKS) {
+				link_iter = rcu_dereference(iter->link[link_id_iter]);
+				if (!link_iter)
+					continue;
+				/* Check if any of link of iterator sdata belongs
+				 * to same mbssid group as the tx link
+				 */
+				if (link_iter->conf->mbssid_tx_vif != vif ||
+				    link_iter->conf->mbssid_tx_vif_linkid != link->link_id)
+					continue;
+
+				ieee80211_queue_work(&iter->local->hw,
+						     &link_iter->csa_finalize_work);
+			}
 		}
 	}
 	ieee80211_queue_work(&local->hw, &link->csa_finalize_work);
@@ -4898,6 +4927,7 @@ ieee80211_color_change_bss_config_notify(struct ieee80211_link_data *link,
 {
 	struct ieee80211_sub_if_data *sdata = link->sdata;
 
+	sdata_assert_lock(sdata);
 	link->conf->he_bss_color.color = color;
 	link->conf->he_bss_color.enabled = enable;
 
@@ -4905,19 +4935,38 @@ ieee80211_color_change_bss_config_notify(struct ieee80211_link_data *link,
 
 	ieee80211_link_info_change_notify(sdata, link, changed);
 
-	/* handle bss color in mbssid case with MLO */
-	if (!link->conf->nontransmitted && sdata->vif.mbssid_tx_vif) {
+	if (!link->conf->nontransmitted && link->conf->mbssid_tx_vif) {
+		unsigned int link_id_iter;
+		u16 valid_links;
 		struct ieee80211_sub_if_data *child;
+		struct ieee80211_link_data *link_iter;
 
 		mutex_lock(&sdata->local->iflist_mtx);
 		list_for_each_entry(child, &sdata->local->interfaces, list) {
-			if (child != sdata && child->vif.mbssid_tx_vif == &sdata->vif) {
-				child->vif.bss_conf.he_bss_color.color = color;
-				child->vif.bss_conf.he_bss_color.enabled = enable;
-				ieee80211_link_info_change_notify(child,
-								  &child->deflink,
+			if (child == sdata)
+				continue;
+			sdata_lock(child);
+			/* check deflink by default */
+			valid_links = child->vif.valid_links | BIT(0);
+			for_each_set_bit(link_id_iter, &valid_links,
+					 IEEE80211_MLD_MAX_NUM_LINKS) {
+				link_iter = sdata_dereference(child->link[link_id_iter],
+							      child);
+				if (!link_iter)
+					continue;
+				/* Check if any of link of iterator sdata belongs
+				 * to same mbssid group as the tx link
+				 */
+				if (link_iter->conf->mbssid_tx_vif != &sdata->vif ||
+				    link_iter->conf->mbssid_tx_vif_linkid != link->link_id)
+					continue;
+
+				link_iter->conf->he_bss_color.color = color;
+				link_iter->conf->he_bss_color.enabled = enable;
+				ieee80211_link_info_change_notify(child, link_iter,
 								  BSS_CHANGED_HE_BSS_COLOR);
 			}
+			sdata_unlock(child);
 		}
 		mutex_unlock(&sdata->local->iflist_mtx);
 	}
diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index c55f6b4..90fad65 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -673,28 +673,36 @@ static void ieee80211_do_stop(struct ieee80211_sub_if_data *sdata, bool going_do
 static void ieee80211_stop_mbssid(struct ieee80211_sub_if_data *sdata)
 {
 	struct ieee80211_sub_if_data *tx_sdata, *non_tx_sdata, *tmp_sdata;
-	struct ieee80211_vif *tx_vif = sdata->vif.mbssid_tx_vif;
+	struct ieee80211_vif *tx_vif = sdata->vif.bss_conf.mbssid_tx_vif;
 
 	if (!tx_vif)
 		return;
 
+	/* TODO: For MLO Need to handle this teardown by looping through all links bss
+	 * in mbssid group, if any of such link is part of MLD Stop all link in that MLD.
+	 * Currenly handled Non-MLO teardown.
+	 */
+	if (tx_vif->active_links)
+		return;
+
 	tx_sdata = vif_to_sdata(tx_vif);
-	sdata->vif.mbssid_tx_vif = NULL;
+	sdata->vif.bss_conf.mbssid_tx_vif = NULL;
 
 	list_for_each_entry_safe(non_tx_sdata, tmp_sdata,
 				 &tx_sdata->local->interfaces, list) {
 		if (non_tx_sdata != sdata && non_tx_sdata != tx_sdata &&
-		    non_tx_sdata->vif.mbssid_tx_vif == tx_vif &&
+		    non_tx_sdata->vif.bss_conf.mbssid_tx_vif == tx_vif &&
 		    ieee80211_sdata_running(non_tx_sdata)) {
-			non_tx_sdata->vif.mbssid_tx_vif = NULL;
+			non_tx_sdata->vif.bss_conf.mbssid_tx_vif = NULL;
 			dev_close(non_tx_sdata->wdev.netdev);
 		}
 	}
 
 	if (sdata != tx_sdata && ieee80211_sdata_running(tx_sdata)) {
-		tx_sdata->vif.mbssid_tx_vif = NULL;
+		tx_sdata->vif.bss_conf.mbssid_tx_vif = NULL;
 		dev_close(tx_sdata->wdev.netdev);
 	}
+	return;
 }
 
 static int ieee80211_stop(struct net_device *dev)
-- 
2.17.1

